include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,413 @@
+
+template <::std::uint16_t __data_per_workitem, ::std::uint16_t __workgroup_size, typename _Type, typename _FlagType,
+          typename _InRng, typename _OutRng, typename _BinaryOp, typename _StatusFlags, typename _StatusValues,
+          typename _TileVals>
+struct __lookback_kernel_func
+{
+    using _FlagStorageType = typename _FlagType::_FlagType;
+    static constexpr std::uint32_t __elems_in_tile = __workgroup_size * __data_per_workitem;
+
+    _InRng __in_rng;
+    _OutRng __out_rng;
+    _BinaryOp __binary_op;
+    ::std::size_t __n;
+    _StatusFlags __status_flags;
+    ::std::size_t __status_flags_size;
+    _StatusValues __status_vals_full;
+    _StatusValues __status_vals_partial;
+    ::std::size_t __current_num_items;
+    _TileVals __tile_vals;
+
+    [[sycl::reqd_sub_group_size(SUBGROUP_SIZE)]] void
+    operator()(const sycl::nd_item<1>& __item) const
+    {
+        auto __group = __item.get_group();
+        auto __subgroup = __item.get_sub_group();
+        auto __local_id = __item.get_local_id(0);
+
+        ::std::uint32_t __tile_id = 0;
+
+        // Obtain unique ID for this work-group that will be used in decoupled lookback
+        if (__group.leader())
+        {
+            sycl::atomic_ref<_FlagStorageType, sycl::memory_order::relaxed, sycl::memory_scope::device,
+                             sycl::access::address_space::global_space>
+                __idx_atomic(__status_flags[__status_flags_size - 1]);
+            __tile_id = __idx_atomic.fetch_add(1);
+        }
+
+        __tile_id = sycl::group_broadcast(__group, __tile_id, 0);
+
+        ::std::size_t __current_offset = static_cast<::std::size_t>(__tile_id) * __elems_in_tile;
+        auto __out_begin = __out_rng.begin() + __current_offset;
+
+        if (__current_offset >= __n)
+            return;
+
+        // Global load into local
+        auto __wg_current_offset = (__tile_id * __elems_in_tile);
+        auto __wg_next_offset = ((__tile_id + 1) * __elems_in_tile);
+        auto __wg_local_memory_size = __elems_in_tile;
+
+        if (__wg_next_offset > __n)
+            __wg_local_memory_size = __n - __wg_current_offset;
+
+        if (__wg_next_offset <= __n)
+        {
+            _ONEDPL_PRAGMA_UNROLL
+            for (std::uint32_t __i = 0; __i < __data_per_workitem; ++__i)
+            {
+                __tile_vals[__local_id + __workgroup_size * __i] =
+                    __in_rng[__wg_current_offset + __local_id + __workgroup_size * __i];
+            }
+        }
+        else
+        {
+            _ONEDPL_PRAGMA_UNROLL
+            for (std::uint32_t __i = 0; __i < __data_per_workitem; ++__i)
+            {
+                if (__wg_current_offset + __local_id + __workgroup_size * __i < __n)
+                {
+                    __tile_vals[__local_id + __workgroup_size * __i] =
+                        __in_rng[__wg_current_offset + __local_id + __workgroup_size * __i];
+                }
+            }
```

Comment:
I'm not sure I follow the suggestion. The outer branch here (`__wg_next_offset <= n`) is giving a measurable performance increase in a small experiment that I tried (around 5% for an input of size 4M).

This branch is only taken for the very last work-group. Are you suggesting to rewrite this loop?

Reply:
I agree that this outer branch is the correct choice.  I'm wondering about what is in the `else`.  

Right now you unroll and check if each index is under `__n` in the innermost loop.
Another option would be to have some logic to calculate for each work-item the number of elements it will do something with, then loop from 0 to that number.  This avoids the branching in the inner-most loop, but at the cost of not-unrolling the loop.

I'm not sure it will be a benefit, and may be a small margin of a small subset of the computation. 

Not a blocker.